### PR TITLE
many: rename "snap-confine" to "snap-wrap" to workaround LP:1673247

### PR DESCRIPTION
**Pull request by mvo5**
There is a dpkg bug triggered in ubiquity (see LP:1673247 for details) when snapd is upgraded to 2.23.1 because the conffile /etc/apparmor.d/usr.lib.snapd.snap-cofine moves from the snap-confine to the snapd package. To workaround we have to rename the binary so that the apparmor file gets a different name no more name clash can happen.

We also need to squash merge this for easier cherry-picking itno the 2.23 branch.

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
That's just gross. Why not just rename the apparmor profile name instead the whole bloody thing?

File: packaging/ubuntu-16.04/control [hunk 1]
```diff
@@ -62,8 +62,8 @@ Depends: adduser,
          systemd,
          ${misc:Depends},
          ${shlibs:Depends}
-Replaces: ubuntu-snappy (<< 1.9), ubuntu-snappy-cli (<< 1.9), snap-confine (<< 2.23), ubuntu-core-launcher (<< 2.22)
-Breaks: ubuntu-snappy (<< 1.9), ubuntu-snappy-cli (<< 1.9), snap-confine (<< 2.23), ubuntu-core-launcher (<< 2.22)
+Replaces: ubuntu-snappy (<< 1.9), ubuntu-snappy-cli (<< 1.9), snap-wrap (<< 2.23), ubuntu-core-launcher (<< 2.22)
```
Comment (zyga) [inline packaging/ubuntu-16.04/control]: Is this intentional? We don't have snap-wrap (ever) as a standalone package and we do want to replace snap-conifne

File: packaging/ubuntu-16.04/control [hunk 2]
```diff
@@ -96,7 +96,7 @@ Section: oldlibs
 Description: transitional dummy package
  This is a transitional dummy package. It can safely be removed.
 
-Package: snap-confine
```
Comment (zyga) [inline packaging/ubuntu-16.04/control]: I think you want to keep snap-confine as-is (empty package)

File: spread.yaml [hunk 3]
```diff
@@ -316,7 +316,7 @@ suites:
         restore: |
             $TESTSLIB/reset.sh --store
             if [[ "$SPREAD_SYSTEM" != ubuntu-core-16-* ]]; then
-                apt-get purge -y snapd snap-confine ubuntu-core-launcher
+                apt-get purge -y snapd snap-wrap ubuntu-core-launcher
```
Comment (zyga) [inline spread.yaml]: Related to comments above

**Comment by mvo5** [comment]
@Conan-Kudo I'm not disagreeing that it is gross :/ Sorry for that. We tried hard to find a better workaround, the dpkg bug is fixed but we can not fix the live-cds with broken dpkg. Renaming the apparmor profile is tricky as it must match the exact path of the binary it confines.

**Comment by Conan-Kudo** [comment]
@mvo5 Two possibly less awful choices:

1. Since the merge of snap-confine into snapd occurred post Xenial and Yakkety, have Breaks statements to force dpkg to get upgraded before handling snap-confine -> snapd.

2. Empty the profile in the snap-confine package and ship a new profile in snapd package.

What prevented either of those from working?

**Comment by adconrad** [comment]
@Conan-Kudo Responding to both of the less awful choices, in reverse order:

2. We did this already.  Due to a dpkg bug when called with --root=/somepath, the conffile comparison was being made against the root directory instead of the target, thus the installer's attempt to upgrade the /target chroot failed due to thinking there was a file overlap that wasn't there.
1. The installer uses the root system's dpkg, a pre-depends (or any package relationship) that forced dpkg to be upgraded in the /target where we're operating would do no good, as we'd still be performing the operation with the broken dpkg on the CD.

So, yes.  dpkg needs fixing for this bug (and that's in progress), but the only way out of the "ISOs connected to the internet opportunistically upgrade packages on the fly, and our upgrade breaks, oops" issue created by the combination of this package shuffle and the dpkg bug is to shuffle differently, ie: with the binary rename.  At the end of the day, it's an internal libexec binary anyway, so from my POV, ugly though the solution is, it's all a bit of a "meh".  You could rename it to george-of-the-jungle and it wouldn't have any practical impact.

File: tests/upgrade/basic/task.yaml [hunk 4]
```diff
@@ -13,7 +13,7 @@ execute: |
     . "$TESTSLIB/apt.sh"
 
     echo Install previous version...
-    dpkg -l snapd snap-confine || true
+    dpkg -l snapd snap-wrap || true
```
Comment (adconrad) [inline tests/upgrade/basic/task.yaml]: This was an accidental sed, as is refers to package names.  Though it's also possible this is obsolete now, and all you care about is snapd?

Comment (mvo5) [inline tests/upgrade/basic/task.yaml]: Thank you!

**Comment by zyga** [comment]
The failing security test needs investigation. I suspect that the rename needs extra work to actually use snap-wrap (because we rely on snap-confine's existing profile in the system)

**Comment by niemeyer** [comment]
@Conan-Kudo This is just an internal tool being renamed. Let's not overvalue the change.

**Comment by jdstrand** [comment]
@mvo5 - You said: "I'm not disagreeing that it is gross :/ Sorry for that. We tried hard to find a better workaround, the dpkg bug is fixed but we can not fix the live-cds with broken dpkg. Renaming the apparmor profile is tricky as it must match the exact path of the binary it confines."

Actually, that is not true. The name of the profile can be anything. Since we are using the binary attachment the *contents* of the file need to be specific to the binary it confines. As such, you can simply do:
```
$ cat /etc/apparmor.d/snap-confine
...
/usr/lib/snapd/snap-confine (attach_disconnected) {
...
```

and it will work perfectly fine.

**Comment by jdstrand** [comment]
Note that if you rename to /etc/apparmor.d/snap-confine and /etc/apparmor.d/usr.lib.snapd.snap-confine is present, /etc/apparmor.d/usr.lib.snapd.snap-confine will end up replacing /etc/apparmor.d/snap-confine during kernel load since it is parsed after /etc/apparmor.d/snap-confine and the attachment is the for the same binary. I'm not super excited about the name change either. If we must name change, how about something that is more appropriate-- snap-wrap doesn't convey everything it does. Perhaps snap-sandbox?

**Comment by niemeyer** [comment]
@jdstrand About the name, nothing will convey everything it does. I'd say snap-wrap is actually more realistic in that it conveys better the fact that this is not simply a sandbox tool. It rather sets up the whole environment for the snap to run under.

That said, if we can opt to not rename it at all by simply renaming that one file, that sounds even better.

**Comment by jdstrand** [comment]
The apparmor package will do `ls -1` in /etc/apparmor.d so to honor sort ordering by `ls`, I suggest using one of the following:
- /etc/apparmor.d/usr.lib.snapd.snap-confine.real
- /etc/apparmor.d/usr.lib.snapd.snap-confine.ubuntu
- /etc/apparmor.d/zz-snap-confine

In this manner, if the old /etc/apparmor.d/usr.lib.snapd.snap-confine happens to remain present for any reason, the new one (eg, /etc/apparmor.d/usr.lib.snapd.snap-confine.real) will always be loaded after and used instead. Obviously, feel free to use another filename for the new profile that makes the name sort after /etc/apparmor.d/usr.lib.snapd.snap-confine.

**Comment by zyga** [comment]
I give up on figuring out what is wrong with the update test. I have opened #3064 as an (easier) alternative solution that @jdstrand pointed out above.

**Comment by adconrad** [comment]
@jdstrand: What "any reason" would it be left lying around?  It's our conffile, and we can very easily ensure it's wiped completely off the system.  If just removing the old conffile and installing a new one with a new name is the easiest path here, I'm very +1 on scrapping this branch and comitting that few-line fix instead.

**Comment by jdstrand** [comment]
@adconrad - it is possible with rm_conffile and older methods to not remove the conffile if the user modified it. Note that https://github.com/snapcore/snapd/pull/3064 is the updated branch. As a member of the SRU team, your eyes on that PR would be appreciated by the snappy folks I imagine.

**Comment by chipaca** [comment]
Closing as it's superseded by #3064

**Pull request closed without merging**